[dnf5] Simplify metadata downloading by using LoadFlags everywhere

dmach opened the pull request.

dmach committed:
Simplify metadata downloading by using LoadFlags everywhere

dmach committed:
Download and load only PRIMARY by default

dmach committed:
Download and load FILELISTS only when a path is in pkg_specs

Apply this only to the selected commands such as install, download,
repoquery.

lukash commented on libdnf/repo/repo_downloader.hpp:
```diff
@@ -26,6 +26,7 @@ along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
 #include "libdnf/base/base_weak.hpp"
 #include "libdnf/common/exception.hpp"
 #include "libdnf/repo/config_repo.hpp"
+#include "libdnf/repo/repo.hpp"
```
What I don't like about this is that it creates a virtual cycle in the class hierarchy. `Repo` uses `RepoDownloader` and now you've added the reverse dependency of `RepoDownloader` using `Repo`(`::LoadFlags`).

I know it works right now, but it could cause issues in the future.

To maintain a uni-directional hierarchy the `LoadFlags` should be split out of the `Repo` class and into its own header file I think. I've been meaning to go in this direction.

dmach replied:
Makes sense to split LoadFlags into its own header file.

Can you also check if the download/load laziness makes sense to you?
In general it reduces both download sizes and startup times (less data loaded into sack).
I see a challenge in tweaking code of individual commands to detect what needs to be downloaded and loaded.

lukash replied:
I wouldn't call it laziness, you just don't download the data you don't need. But on the next run, with a different command, more data can be needed and it just won't be there. I don't see how we can do this unless we check what metadata are actually downloaded and download anything else that's needed on-demand.

Another "issue" would be automatic downloading metadata in the background, what would that download? All of them?

dmach replied:
> Another "issue" would be automatic downloading metadata in the background, what would that download? All of them?

All metadata for dependency resolution and applying updateinfo should be downloaded by default.
This means: primary, modulemd, updateinfo, maybe also comps if we're going to support upgrading groups.
In other words: filelists and others (== changelogs) are not required in most cases.

lukash replied:
Ok. I guess that's a detail, the list of metadata to download for automatic can be either hardcoded or even eventually configurable.

You didn't address my first point, which is the real issue here AFAICS. Seems that'll be harder to tackle, how about removing the "lazy" loading from the PR and keep the downloaded metadata set the same (and just separate the `LoadFlags` into its own header). I'd merge that and you can post another PR for the loading if/when you want to sort out the "additional" downloading issue.

We'd prefer PRs to be merged instead of hanging around indefinitely :grin:

dmach replied:
I'll make the changes incl. `LoadFlags` refactoring as soon as time allows.
Thanks for your feedback.

dmach committed:
Explicitly load PRIMARY when COMPS gets loaded

PRIMARY must always be loaded otherwise Repo::load_available_repo() errors out.

dmach commented:
I don't plan any more changes to this PR, it is good to go from my perspective.
It has also passed the existing CI tests - even if `FILELISTS` are not loaded in many cases.

lukash commented:
Thanks!
